r/R/duckdb.R

```diff
@@ -0,0 +1,117 @@
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+#' Create a (virtual) DuckDB table from an Arrow object
+#'
+#' This will do the necessary configuration to create a (virtual) table in DuckDB
+#' that is backed by the Arrow object given. No data is copied or modified until
+#' `collect()` or `compute()` are called or a query is run against the table.
+#'
+#' The result is a dbplyr-compatible object that can be used in d(b)plyr pipelines.
+#'
+#' @param src the Arrow object (e.g. Dataset, Table) to use for the DuckDB table
+#' @param con a DuckDB connection to use (default will create one and store it
+#' in `options("arrow_duck_con")`)
+#' @param table_name a name to use in DuckDB for this object. The default is a
+#' unique string `"arrow_"` followed by numbers.
+#' @param auto_disconnect should the table be automatically cleaned up when the
+#' resulting object is removed (and garbage collected)? Default: `TRUE`
+#'
+#' @return A tbl of the new table in DuckDB
+#'
+#' @keywords internal
+#' @name to_duckdb
+NULL
+
+arrow_duck_connection <- function() {
+  con <- getOption("arrow_duck_con")
+  if (is.null(con) || !DBI::dbIsValid(con)) {
+    con <- DBI::dbConnect(duckdb::duckdb())
+    # Use the same CPU count that the arrow library is set to
+    DBI::dbExecute(con, paste0("PRAGMA threads=", cpu_count()))
+    options(arrow_duck_con = con)
+  }
+  con
+}
+
+# Adapted from dbplyr
+unique_arrow_tablename <- function () {
+  i <- getOption("arrow_table_name", 0) + 1
+  options(arrow_table_name = i)
+  sprintf("arrow_%03i", i)
+}
+
+.alchemize_to_duckdb <- function(
+  src,
+  con = arrow_duck_connection(),
+  table_name =  unique_arrow_tablename(),
+  auto_disconnect = TRUE) {
+  duckdb::duckdb_register_arrow(con, table_name, src)
+
+  tbl <- tbl(con, table_name)
+  groups <- src$group
+  if (length(groups) > 0 && !is.null(groups)) {
+    tbl <- dplyr::group_by(tbl, !!rlang::sym(groups))
+  }
+
+  if (auto_disconnect) {
+    # this will add the correct connection disconnection when the tbl is gced.
+    # we should probably confirm that this use of src$disco is kosher.
+    tbl$src$disco <- duckdb_disconnector(con, table_name)
+  }
+
+  tbl
+}
+
+#' @rdname to_duckdb
+#' @export
+to_duckdb <- function(src, con, table_name, auto_disconnect) UseMethod("to_duckdb", src)
+
+#' @rdname to_duckdb
+#' @export
+to_duckdb.Dataset <- .alchemize_to_duckdb
+
+#' @rdname to_duckdb
+#' @export
+to_duckdb.Table <- .alchemize_to_duckdb
+
+#' @rdname to_duckdb
+#' @export
+to_duckdb.arrow_dplyr_query <- .alchemize_to_duckdb
+
+summarise_duck <- function(.data, ...) {
+  # TODO: pass a connection?
```

Comment:
I don't think so--if the user wants more control over the duckdb connection, they can call `to_duckdb()` explicitly rather than implicitly here.